packages/core/src/typeschema/crawler.ts

```diff
@@ -0,0 +1,113 @@
+import { Resource } from '@medplum/fhirtypes';
+import { getTypedPropertyValue, toTypedValue } from '../fhirpath';
+import { TypedValue } from '../types';
+import { arrayify, isLowerCase } from '../utils';
+import { getDataType, InternalTypeSchema } from './types';
+
+export interface ResourceVisitor {
+  onEnterObject?: (path: string, value: TypedValue, schema: InternalTypeSchema) => void;
+  onExitObject?: (path: string, value: TypedValue, schema: InternalTypeSchema) => void;
+  visitProperty?: (
+    parent: TypedValue,
+    key: string,
+    path: string,
+    propertyValues: (TypedValue | TypedValue[] | undefined)[],
```

Comment:
The nested array is only necessary when retrieving data from a deep path: if this is only looking at a single direct property of one object in the JSON, this could probably be `TypedValue[] | undefined` or similar.  Maybe worth splitting out the common case from the weird profile case where nested paths are used — perhaps `visitProperty` and `visitDeepPath`?

Reply:
If ok with you, this is something I'd like to revisit in future.  I don't fully understand the cases where we need nested arrays.

I noticed that the case where evaluating an `ElementDefinition` without a value is still allocating empty arrays, which seems like low hanging fruit for optimizations.

I assume this was done to handle the slicing case?

I'm comfortable leaving as-is for now, and revisiting as we find more use cases for the crawler.